### PR TITLE
Add Iterator, and optimize equals & intersection methods to reduce iterations

### PR DESCRIPTION
**Pull request by EricPKerr**

### CODE REVIEW

**Comment by jpbulman** [comment]
@EricPKerr That would definitely be cool to add an iterator feature, if you want to take that up!

**Comment by EricPKerr** [comment]
Cool, updated and added tests

File: src/AdvancedSet.ts [hunk 1]
```diff
@@ -43,7 +43,15 @@ export default class AdvancedSet<T> {
    * Returns a new set containing the shared elements between this set and the given parameter set
    */
   public intersection(setB: AdvancedSet<T>): AdvancedSet<T> {
-    return new AdvancedSet(...this.toArray().filter((x) => setB.has(x)));
+    let smallest: AdvancedSet<T> = setB;
+    let other: AdvancedSet<T> = this;
+    
+    if (this.size < setB.size) {
```
Comment (jpbulman) [inline src/AdvancedSet.ts:49]: This is a great approach!

File: src/AdvancedSet.ts [hunk 2]
```diff
@@ -120,6 +132,18 @@ export default class AdvancedSet<T> {
     return this.intersection(setB).isEmpty();
   }
 
+  [Symbol.iterator]() {
+    let index = -1;
+    let data = this.toArray();
+
+    return {
+      next: () => ({
```
Comment (jpbulman) [inline src/AdvancedSet.ts:140]: Awesome :rocket: !

**Comment by jpbulman** [comment]
LGTM! Thanks for doing this @EricPKerr! 

Resolves #8

**Pull request merged**